go/arrow/array/dictionary.go

```diff
@@ -1674,6 +1674,135 @@ func (u *unifier) GetResultWithIndexType(indexType arrow.DataType) (arrow.Array,
 	return MakeFromData(dictData), nil
 }
 
+type binaryUnifier struct {
+	mem       memory.Allocator
+	memoTable *hashing.BinaryMemoTable
+}
+
+// NewBinaryDictionaryUnifier constructs and returns a new dictionary unifier for dictionaries
+// of binary values, using the provided allocator for allocating the unified dictionary
+// and the memotable used for building it.
+func NewBinaryDictionaryUnifier(alloc memory.Allocator) DictionaryUnifier {
+	return &binaryUnifier{
+		mem:       alloc,
+		memoTable: hashing.NewBinaryMemoTable(0, 0, NewBinaryBuilder(alloc, arrow.BinaryTypes.Binary)),
+	}
+}
+
+func (u *binaryUnifier) Release() {
+	u.memoTable.Release()
+}
+
+func (u *binaryUnifier) Unify(dict arrow.Array) (err error) {
+	if !arrow.TypeEqual(arrow.BinaryTypes.Binary, dict.DataType()) {
+		return fmt.Errorf("dictionary type different from unifier: %s, expected: %s", dict.DataType(), arrow.BinaryTypes.Binary)
+	}
```

Comment:
Since we're going to require a binary dictionary, any reason to not just take in a `*array.Binary` and push the type check onto the consumer?

Reply:
I was just following the `DictionaryUnifier` interface. If we want to do that, we should export the `binaryUnifier` struct. Happy to go with either. Also happy to add additional functions that are type safe in addition.

Reply:
Ah! I completely forgot that I had the exported interface. My bad there. I think it's better to follow the interface for now probably.